toggle debug reporting via CPP

AndrasKovacs opened the pull request:
Disable generating any code for reporting functions in Agda.TypeChecking.Monad.Debug if the DEBUG flag is not enabled. This yields noticable build time and runtime speedups.

Previously the DEBUG flag was only used in serialization. That usage is now relegated to the new DEBUG_SERIALISATION flag.

There are two new functions, `alwaysReportSLn` and `alwaysReportSDoc`, which are always active, regardless of the DEBUG flag. These are used wherever we previously printed with verbosity 1, so default printing output is unchanged.

AndrasKovacs commented:
Agda build times on my machine:

- no heavy opt, debug: 7m3,187s
- no heavy opt,  no debug: 6m13,612s
- heavy opt, debug: 11m57,993s
- heavy opt, no debug: 10m13,333s

Stdlib build times:
- heavy opt, debug:     6m8,176s
- heavy opt, no debug:   5m45,000s
- no heavy opt, debug:   9m43,940s
- no heavy opt, no debug: 8m47,484s

AndrasKovacs commented:
It seems that all test failures are because of verbose output being included in golden tests.

nad commented:
> Previously the DEBUG flag was only used in serialization.

It is also used by the operator parser:

https://github.com/agda/agda/blob/9e6776225a0ec19ce0cb4b047cc3b6c8a2a826c4/src/full/Agda/Syntax/Concrete/Operators/Parser/Monad.hs#L43-L51

AndrasKovacs commented:
@nad You're right, I missed it. I think it makes sense to have a separate flag for that as well.

What about the golden tests containing verbose output, can I simply update them?  There are 29 such tests overall. I list them:

```
test/Bugs/Issue3898.err
test/Compiler/simple/CaseOnCase.out
test/Compiler/simple/CompareNat.out
test/Compiler/simple/CompileCatchAll.out
test/Compiler/simple/CompileNumbers.out
test/Compiler/simple/EraseRefl.out
test/Compiler/simple/InlineRecursive.out
test/Compiler/simple/UnusedArguments.out
test/Compiler/simple/Word.out
test/Fail/ConstructorHeadedPointlessForRecordPatterns.err
test/Fail/DebugWith.err
test/Fail/Issue1963DisplayWithPostfixCopattern.err
test/Fail/Issue2018debug.err
test/Fail/Issue3590-1.err
test/Fail/Issue3590-2.err
test/Fail/Issue4175.err
test/Fail/Issue641-all-interfaces.err
test/Fail/Issue641.err
test/Fail/Optimised-open.err
test/Succeed/ForcedConstructorPattern.warn
test/Succeed/Issue2571.warn
test/Succeed/Issue2908.warn
test/Succeed/RepeatedCase.warn
test/interaction/Debug.out
test/interaction/Issue1353.out
test/interaction/Issue4917.out
test/interaction/Issue6807.out
test/interaction/Positivity-once.outside
```
Plus there's the `examples/uptodate` which looks for a specific `v=5` message in the output.

AndrasKovacs commented:
- I added a new DEBUG_PARSING flag to cover the previous use-case in parsing. 
- I mentioned DEBUG in the changelog.
- I added the debug flag to the `quicker-install-bin` build.

Can I just update the golden tests?

andreasabel commented:
Links to golden values using debug output:
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/Bugs/Issue3898.err
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/Compiler/simple/CaseOnCase.out
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/Compiler/simple/CompareNat.out
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/Compiler/simple/CompileCatchAll.out
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/Compiler/simple/CompileNumbers.out
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/Compiler/simple/EraseRefl.out
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/Compiler/simple/InlineRecursive.out
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/Compiler/simple/UnusedArguments.out
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/Compiler/simple/Word.out
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/Fail/ConstructorHeadedPointlessForRecordPatterns.err
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/Fail/DebugWith.err
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/Fail/Issue1963DisplayWithPostfixCopattern.err
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/Fail/Issue2018debug.err
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/Fail/Issue3590-1.err
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/Fail/Issue3590-2.err
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/Fail/Issue4175.err
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/Fail/Issue641-all-interfaces.err
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/Fail/Issue641.err
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/Fail/Optimised-open.err
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/Succeed/ForcedConstructorPattern.warn
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/Succeed/Issue2571.warn
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/Succeed/Issue2908.warn
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/Succeed/RepeatedCase.warn
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/interaction/Debug.out
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/interaction/Issue1353.out
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/interaction/Issue4917.out
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/interaction/Issue6807.out
- https://github.com/agda/agda/blob/86fce9a304da38e44175ad2478bc77893d2ea5e6/test/interaction/Positivity-once.outside

AndrasKovacs committed:
Cover verboseBracket with the debug flag

AndrasKovacs committed:
Run tests in CI with -fdebug

Add a new install-bin-debug target to the makefile, which
enables the debug flag. Use that in the CI test build
of Agda.

Also add the new debug flags to stack-dry-run.yml
as NON_DEFAULT_FLAGS.

AndrasKovacs committed:
toggle debug reporting via CPP

Disable generating any code for reporting functions in Agda.TypeChecking.Monad.Debug
if the DEBUG flag is not enabled. This yields noticable build time and runtime
speedups.

Previously the DEBUG flag only had minor use in serialization and parsing.
Those usages are now relegated to the new DEBUG_SERIALISATION and DEBUG_PARSING
flags.

There are two new functions, `alwaysReportSLn` and `alwaysReportSDoc`, which are
always active, regardless of the DEBUG flag. These are used wherever we
previously printed with verbosity 1, so default printing output is unchanged.

AndrasKovacs commented:
It seems that cluster counting is not enabled in the CI build and that's why `latex-html-test` fails there. On my machine it works fine, and I don't really understand why the CI differs. My latest commit added an `install-bin-debug` make target, which is the same as `install-bin` except with `-fdebug`, and used that in the build phase of `test.yml`.

andreasabel reviewed (approved):
Please also add the `-f debug` to the developer goals 
- [x] `quicker-*`
- [x] `fast-*`
- [x]  `v1-install`

Welcome to add more variables to the `make debug` target.

AndrasKovacs committed:
Fix Makefile, add debug to dev targets, more "make debug" printing
